src/boosting/sample_strategy.cpp

```diff
@@ -0,0 +1,25 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+
+#include <LightGBM/sample_strategy.h>
+#include "goss.hpp"
+#include "bagging.hpp"
+
+namespace LightGBM {
+
+SampleStrategy* SampleStrategy::CreateSampleStrategy(const Config* config, const Dataset* train_data, const ObjectiveFunction* objective_function, int num_tree_per_iteration) {
+  bool use_goss_as_boosting = config->boosting == std::string("goss");
+  bool use_goss_as_strategy = config->data_sample_strategy == std::string("goss");
+  if (use_goss_as_boosting) {
```

Comment:
`goss` is still listed as a valid value for `boosting`  in `config.h` / `Parameters.rst`.

Could you please update that documentation in this PR with information similar to the message in this warning? So users will know that `boosting=goss` is  only supported for backwards compatibility, but that it isn't recommended.

Reply:
I believe we should simply remove `goss` option for `boosting` parameter in new version of documentation. Newcomers will use `data_sample_strategy=goss` from their first steps with LightGBM while already existing code with `boosting=goss` will still work but with appropriate warning.

Reply:
I'm ok with removing it completely from the documentation, instead. Let's do that.

I was torn between these two competing concerns:

* keep it (*"the documentation should match the state of the library"*)
* remove it (*"discouraged or deprecated patterns should be hidden"*)

Reply:
Ok. I'll remove that.

Reply:
Removed via cddfcd6

Reply:
Also, please remove `goss` from these places:
https://github.com/microsoft/LightGBM/blob/5631366175f68a28bd174982cca16975f7fb48af/python-package/lightgbm/sklearn.py#L376
https://github.com/microsoft/LightGBM/blob/df79713725649d5177fd0974f158b20a1b57e971/docs/Development-Guide.rst#important-classes
"... or ``goss`` boosting ..." -> "... or ``goss`` data sample strategy ..."
https://github.com/microsoft/LightGBM/blob/6ced58ad9aade8200cb592c779581050b7102b19/include/LightGBM/config.h#L250

Reply:
Resolved via 4bdcdd5